zerver/views/streams.py

```diff
@@ -689,11 +691,10 @@ def add_subscriptions_backend(
         user_profile, existing_streams
     )
     if len(unauthorized_streams) > 0 and authorization_errors_fatal:
-        raise JsonableError(
-            _("Unable to access channel ({channel_name}).").format(
-                channel_name=unauthorized_streams[0].name,
-            )
-        )
+        if unauthorized_streams[0].deactivated:
+            raise CannotAccessArchivedStreamError(unauthorized_streams[0].name)
+        raise CannotAccessPrivateStreamError(unauthorized_streams[0].name)
```

Comment:
Maybe a comment here mentioning that we know stream is private because non-guest users can access all public streams and guest users are already handled by `require_non_guest_user` would be good.